moveit_planners/trajopt/src/trajopt_interface.cpp

```diff
@@ -0,0 +1,458 @@
+#include <trajopt_sco/solver_interface.hpp>
+
+#include <trajopt/trajectory_costs.hpp>
+
+#include <ros/ros.h>
+#include <rosparam_shortcuts/rosparam_shortcuts.h>
+
+#include <limits>
+#include <vector>
+#include <Eigen/Geometry>
+#include <unordered_map>
+
+#include "trajopt_interface/trajopt_interface.h"
+#include "trajopt_interface/problem_description.h"
+
+namespace trajopt_interface
+{
+TrajOptInterface::TrajOptInterface(const ros::NodeHandle& nh) : nh_(nh), name_("TrajOptInterface")
+{
+  prob_ = TrajOptProblemPtr(new TrajOptProblem);
+  setDefaultTrajOPtParams();
+
+  // TODO: callbacks should be defined by the user
+  callbacks_.push_back(callBackFunc);
+}
+
+bool TrajOptInterface::solve(const planning_scene::PlanningSceneConstPtr& planning_scene,
+                             const planning_interface::MotionPlanRequest& req,
+                             moveit_msgs::MotionPlanDetailedResponse& res)
+{
+  ROS_INFO(" ======================================= From trajopt_interface, solve is called");
+  setTrajOptParams(params_);
+
+  if (!planning_scene)
+  {
+    ROS_ERROR_STREAM_NAMED("trajopt_planner", "No planning scene initialized.");
+    res.error_code.val = moveit_msgs::MoveItErrorCodes::FAILURE;
+    return false;
+  }
+
+  if (req.start_state.joint_state.position.empty())
+  {
+    ROS_ERROR_STREAM_NAMED("trajopt_planner", "Start state position is empty");
+    res.error_code.val = moveit_msgs::MoveItErrorCodes::INVALID_ROBOT_STATE;
+    return false;
+  }
+
+  if (req.start_state.joint_state.name.empty())
+  {
+    ROS_ERROR_STREAM_NAMED("trajopt_planner", "Start_state name is empty");
+    res.error_code.val = moveit_msgs::MoveItErrorCodes::INVALID_ROBOT_STATE;
+    return false;
+  }
+
+  ROS_INFO(" ======================================= Extract current state information");
+  ros::WallTime start_time = ros::WallTime::now();
+  robot_model::RobotModelConstPtr robot_model = planning_scene->getRobotModel();
+  robot_state::RobotStatePtr current_state(new robot_state::RobotState(robot_model));
+  *current_state = planning_scene->getCurrentState();
+  const robot_state::JointModelGroup* joint_model_group = current_state->getJointModelGroup(req.group_name);
+  std::vector<std::string> joint_names = joint_model_group->getVariableNames();
+  int dof = joint_names.size();
+  std::vector<double> current_joint_values;
+  current_state->copyJointGroupPositions(joint_model_group, current_joint_values);
+
+  // current state is different from star state in general
+  ROS_INFO(" ======================================= Extract start stae infromation");
+  std::unordered_map<std::string, double> all_joints;
+  trajopt::DblVec joint_start_constraints;
+
+  for (int joint_index = 0; joint_index < req.start_state.joint_state.position.size(); ++joint_index)
+  {
+    all_joints[req.start_state.joint_state.name[joint_index]] = req.start_state.joint_state.position[joint_index];
+  }
+
+  for (auto joint_name : joint_names)
```

Comment:
For robots with mimic joints this probably fails since `all_joints` (from `req.start_state`) usually only contains active joints. I would suggest to test if names and positions of the start state have equal size and then just iterate over those. Inside the loop we could check if the joint name actually refers to an active joint.

Reply:
If we want to check for joints to be active in the loop why do we even want to include the mimic ones in the first place (we are ignoring them with this check)?